homeassistant/helpers/script.py

```diff
@@ -297,22 +297,34 @@ async def _async_event_step(self):
             self._action[CONF_EVENT], event_data, context=self._context
         )
 
-    async def _async_condition_step(self):
-        """Test if condition is matching."""
-        config_cache_key = frozenset((k, str(v)) for k, v in self._action.items())
+    async def _check_condition(self, cond):
+        config_cache_key = frozenset((k, str(v)) for k, v in cond.items())
         config = self._config_cache.get(config_cache_key)
         if not config:
-            config = await condition.async_from_config(self._hass, self._action, False)
+            config = await condition.async_from_config(self._hass, cond, False)
             self._config_cache[config_cache_key] = config
 
+        return config(self._hass, self._variables)
+
+    async def _async_condition_step(self):
+        """Test if condition is matching."""
         self._script.last_action = self._action.get(
             CONF_ALIAS, self._action[CONF_CONDITION]
         )
-        check = config(self._hass, self._variables)
+        check = await self._check_condition(self._action)
         self._log("Test condition %s: %s", self._script.last_action, check)
         if not check:
             raise _StopScript
 
+    async def _async_branch_step(self):
+        branches = self._action.get("branch", [])
+        for b in branches:
+            condition = b.get("if")
+            if not condition or await self._check_condition(condition):
+                for _, self._action in enumerate(b.get("sequence", {})):
+                    await self._async_step(log_exceptions=False)
```

Comment:
I don't think that these should share the `self._async_step` with the parent script. These should be script objects that we call `async_run` on with the same variables

Reply:
Great. That also makes it automagically work with legacy and non-legacy scripts.